File: tiktok_downloader.py
import os
import re
import requests

def sanitize_filename(filename):
    """Membersihkan nama file dari karakter yang tidak diizinkan"""
    filename = re.sub(r'[<>:"/\\|?*]', '', filename)
    filename = filename.strip()
    return filename[:200]  # Batasi panjang nama file

def download_tiktok_mp3(url, output_folder="profile_music"):
    """
    Download audio dari video TikTok dan simpan sebagai MP3
    
    Args:
        url: URL video TikTok
        output_folder: Folder tempat menyimpan file (default: "profile_music")
    """
    
    # Buat folder output jika belum ada
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)
    
    try:
        print(f"Memproses URL: {url}")
        print("Mengambil data dari TikTok...")
        
        # Gunakan API tikwm.com (gratis dan tidak perlu API key)
        api_url = "https://www.tikwm.com/api/"
        
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
        
        # Request ke API
        response = requests.post(
            api_url, 
            data={'url': url, 'hd': 1},
            headers=headers,
            timeout=30
        )
        
        if response.status_code == 200:
            data = response.json()
            
            if data.get('code') == 0:
                # Ambil URL audio
                audio_url = data['data'].get('music')
                
                # Ambil username
                username = data['data']['author'].get('unique_id', 'Unknown')
                
                if audio_url and username:
                    print(f"Username: @{username}")
                    print(f"Mengunduh audio...")
                    
                    # Download audio
                    audio_response = requests.get(audio_url, stream=True, headers=headers)
                    
                    if audio_response.status_code == 200:
                        # Buat nama file
                        safe_username = sanitize_filename(username)
                        filename = f"TiktokSound - {safe_username}.mp3"
                        filepath = os.path.join(output_folder, filename)
                        
                        # Simpan file
                        total_size = int(audio_response.headers.get('content-length', 0))
                        downloaded = 0
                        
                        with open(filepath, 'wb') as f:
                            for chunk in audio_response.iter_content(chunk_size=8192):
                                if chunk:
                                    f.write(chunk)
                                    downloaded += len(chunk)
                                    
                                    # Progress bar sederhana
                                    if total_size > 0:
                                        progress = (downloaded / total_size) * 100
                                        print(f"\rProgress: {progress:.1f}%", end='')
                        
                        print(f"\n✓ Berhasil! File disimpan: {filepath}")
                        print(f"Ukuran file: {os.path.getsize(filepath) / 1024:.2f} KB")
                        return True
                    else:
                        print(f"Gagal mengunduh audio. Status code: {audio_response.status_code}")
                        return False
                else:
                    print("Tidak dapat menemukan URL audio atau username")
                    return False
            elif data.get('code') == -1:
                print("Error: URL TikTok tidak valid atau video tidak ditemukan")
                return False
            else:
                print(f"Error dari API: {data.get('msg', 'Unknown error')}")
                return False
        else:
            print(f"Gagal mengakses API. Status code: {response.status_code}")
            return False
            
    except requests.exceptions.Timeout:
        print("Error: Request timeout. Koneksi internet lambat atau server tidak merespon")
        return False
    except requests.exceptions.RequestException as e:
        print(f"Error koneksi: {str(e)}")
        return False
    except Exception as e:
        print(f"Error: {str(e)}")
        return False

def main():
    """Fungsi utama"""
    print("=" * 50)
    print("TikTok MP3 Downloader")
    print("=" * 50)
    
    # Input URL
    tiktok_url = input("\nMasukkan URL TikTok: ").strip()
    
    # Validasi URL
    if not tiktok_url or 'tiktok.com' not in tiktok_url:
        print("URL TikTok tidak valid!")
        return
    
    # Download
    success = download_tiktok_mp3(tiktok_url)
    
    if success:
        print("\n✓ Download selesai!")
    else:
        print("\n✗ Download gagal!")
    
    input("\nTekan Enter untuk keluar...")

if __name__ == "__main__":
    main()

File: src/server/media/tiktok-downloader.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

/**
 * TikTok Downloader Module untuk Node.js
 * Download audio dari video TikTok menggunakan API tikwm.com
 */

class TikTokDownloader {
    constructor(outputDir = 'profile_music') {
        this.outputDir = outputDir;
        this.apiUrl = 'https://www.tikwm.com/api/';
    }

    /**
     * Membersihkan nama file dari karakter yang tidak diizinkan
     */
    sanitizeFilename(filename) {
        let sanitized = filename.replace(/[<>:"/\\|?*]/g, '');
        sanitized = sanitized.trim();
        return sanitized.slice(0, 200);
    }

    /**
     * Download file dari URL
     */
    async downloadFile(url, filePath) {
        return new Promise((resolve, reject) => {
            const protocol = url.startsWith('https') ? https : http;
            
            const file = fs.createWriteStream(filePath);
            
            protocol.get(url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                }
            }, (response) => {
                if (response.statusCode !== 200) {
                    reject(new Error(`HTTP ${response.statusCode}`));
                    return;
                }

                const totalSize = parseInt(response.headers['content-length'], 10);
                let downloadedSize = 0;

                response.on('data', (chunk) => {
                    downloadedSize += chunk.length;
                    const progress = ((downloadedSize / totalSize) * 100).toFixed(1);
                    console.log(`Progress: ${progress}%`);
                });

                response.pipe(file);

                file.on('finish', () => {
                    file.close();
                    resolve();
                });

                file.on('error', (err) => {
                    fs.unlink(filePath, () => {});
                    reject(err);
                });
            }).on('error', (err) => {
                fs.unlink(filePath, () => {});
                reject(err);
            });
        });
    }

    /**
     * Fetch data dari TikTok API
     */
    async fetchTikTokData(url) {
        return new Promise((resolve, reject) => {
            const postData = `url=${encodeURIComponent(url)}&hd=1`;

            const options = {
                hostname: 'www.tikwm.com',
                path: '/api/',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    'Content-Length': postData.length,
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                },
                timeout: 30000
            };

            const req = https.request(options, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    try {
                        const jsonData = JSON.parse(data);
                        resolve(jsonData);
                    } catch (err) {
                        reject(new Error('Invalid JSON response from API'));
                    }
                });
            });

            req.on('error', reject);
            req.on('timeout', () => {
                req.abort();
                reject(new Error('Request timeout'));
            });

            req.write(postData);
            req.end();
        });
    }

    /**
     * Download TikTok audio
     */
    async downloadTikTokAudio(url) {
        return new Promise(async (resolve, reject) => {
            try {
                // Buat folder output jika belum ada
                if (!fs.existsSync(this.outputDir)) {
                    fs.mkdirSync(this.outputDir, { recursive: true });
                    console.log(`✓ Folder '${this.outputDir}' dibuat`);
                }

                console.log(`Memproses URL: ${url}`);
                console.log('Mengambil data dari TikTok...');

                // Fetch data dari API
                const data = await this.fetchTikTokData(url);

                if (data.code === 0) {
                    const audioUrl = data.data?.music;
                    const username = data.data?.author?.unique_id || 'Unknown';

                    if (!audioUrl) {
                        throw new Error('Tidak dapat menemukan URL audio');
                    }

                    console.log(`Username: @${username}`);
                    console.log('Mengunduh audio...');

                    // Buat nama file
                    const safeUsername = this.sanitizeFilename(username);
                    const filename = `TiktokSound - ${safeUsername}.mp3`;
                    const filepath = path.join(this.outputDir, filename);

                    // Download file
                    await this.downloadFile(audioUrl, filepath);

                    const stats = fs.statSync(filepath);
                    const fileSizeMB = (stats.size / 1024 / 1024).toFixed(2);

                    console.log(`\n✓ Berhasil! File disimpan: ${filename}`);
                    console.log(`Ukuran file: ${fileSizeMB} MB`);

                    resolve({
                        success: true,
                        message: 'Download completed successfully',
                        filename: filename,
                        filepath: filepath,
                        fileSize: stats.size,
                        fileSizeMB: fileSizeMB,
                        username: username
                    });
                } else if (data.code === -1) {
                    throw new Error('URL TikTok tidak valid atau video tidak ditemukan');
                } else {
                    throw new Error(`Error dari API: ${data.msg || 'Unknown error'}`);
                }
            } catch (error) {
                reject(error);
            }
        });
    }

    /**
     * Get list file MP3 di output directory
     */
    async getDownloadedTracks() {
        try {
            const files = fs.readdirSync(this.outputDir);
            const mp3Files = files
                .filter(file => file.endsWith('.mp3'))
                .map(file => ({
                    filename: file,
                    path: path.join(this.outputDir, file),
                    url: `/profile_music/${file}`
                }));
            return mp3Files;
        } catch (error) {
            console.error('Error reading downloads:', error);
            return [];
        }
    }

    /**
     * Delete file track
     */
    async deleteTrack(filename) {
        try {
            const filePath = path.join(this.outputDir, filename);
            if (fs.existsSync(filePath)) {
                fs.unlinkSync(filePath);
                return { success: true, message: 'File deleted successfully' };
            }
            return { success: false, message: 'File not found' };
        } catch (error) {
            console.error('Error deleting file:', error);
            return { success: false, message: error.message };
        }
    }

    /**
     * Get file info
     */
    async getFileInfo(filename) {
        try {
            const filePath = path.join(this.outputDir, filename);
            const stats = fs.statSync(filePath);
            return {
                filename,
                size: stats.size,
                sizeInMB: (stats.size / (1024 * 1024)).toFixed(2),
                createdAt: stats.birthtime,
                modifiedAt: stats.mtime
            };
        } catch (error) {
            console.error('Error getting file info:', error);
            return null;
        }
    }
}

module.exports = TikTokDownloader;

// CLI usage
if (require.main === module) {
    const args = process.argv.slice(2);
    
    if (args.length === 0) {
        console.log("=" .repeat(50));
        console.log("🎵  TIKTOK DOWNLOADER");
        console.log("=" .repeat(50));
        console.log();
        console.log("Usage:");
        console.log("  node tiktok-downloader.js <tiktok_url> [output_dir]");
        console.log();
        console.log("Example:");
        console.log("  node tiktok-downloader.js 'https://www.tiktok.com/@user/video/...' profile_music");
        process.exit(1);
    }

    const tiktokUrl = args[0];
    const outputDir = args[1] || "profile_music";

    console.log("=" .repeat(50));
    console.log("🎵  TIKTOK DOWNLOADER");
    console.log("=" .repeat(50));
    console.log();

    const downloader = new TikTokDownloader(outputDir);
    downloader.downloadTikTokAudio(tiktokUrl)
        .then(result => {
            console.log('\n✅ Download selesai!');
            process.exit(0);
        })
        .catch(error => {
            console.error('\n❌ Download gagal:', error.message);
            process.exit(1);
        });
}
